language-support/ts/daml-ledger/index.ts

```diff
@@ -329,6 +331,277 @@ type LedgerOptions = {
   reconnectThreshold?: number;
 }
 
+class StreamEventEmitter<T extends object, K, I extends string, State> extends EventEmitter implements Stream<T, K, I, State> {
+
+    /** Passed at construction time, called _before_ the 'close' event is emitted */
+    private readonly beforeClosing: () => void;
+
+    constructor({beforeClosing}: { beforeClosing: () => void }) {
+        super();
+        this.beforeClosing = beforeClosing;
+    }
+
+    close(): void {
+        this.beforeClosing();
+        this.emit('close', { code: 4000, reason: "called .close()" });
+        this.removeAllListeners();
+    }
+
+}
+
+type QueryResponseStream<T extends object, K, I extends string> = StreamEventEmitter<T, K, I, readonly CreateEvent<T, K, I>[]>;
+
+type StreamingQuery<T extends object, K, I extends string> = {
+    template: Template<T, K, I>,
+    queries: Query<T>[],
+    stream: QueryResponseStream<T, K, I>,
+    offset?: string | null, // undefined -> nothing has been received, null -> the JSON API returned a null offset
+    state: Map<ContractId<T>, CreateEvent<T, K, I>>, // all JavaScript Map iterators preserve insertion order
+}
+
+type StreamingQueryRequest = {
+    templateIds: string[],
+    query?: object,
+    offset?: string,
+};
+
+function append<K, V>(map: Map<K, V[]>, key: K, value: V): void {
+    if (map.has(key)) {
+        map.get(key)!.push(value);
+    } else {
+        map.set(key, [value]);
+    }
+}
+
+/**
+ * @deprecated All usages of this function should be replaced by just
+ *             iterating over the iterator. For this to happen, the TS
+ *             compiler requires the --downlevelIteration flag, which
+ *             however does not play nicely with Jest when running the
+ *             tests.
```

Comment:
Another option would be to target es6, I think. A quick look at [can I use?](https://caniuse.com/?search=es6) seems to say the last major browser to offer compatibility was Firefox in June 2017. Do we still have a good reason to target es5?

Reply:
No idea. @cocreature?

Reply:
not sure, it’s hard to verify if any of our users rely on es5

Reply:
I'll leave this as is for the time being. No rush to take a decision. I would prefer to keep working against ES5 unless there are strong reasons not to. But honestly "we have been in the third decade of the third millennium for a while" is probably at least a good reason.

Reply:
Embedded the idea as a `TODO` in the code in https://github.com/digital-asset/daml/pull/10221/commits/2cd0bc5ec99961cb088bd74516a1199a2a60e328